test/tests/unit/sessions_tests.js

```diff
@@ -0,0 +1,253 @@
+
+      co(function*() {
+        const mockServer = yield mock.createServer(37019, 'localhost');
+        mockServer.setMessageHandler(request => {
+          request.reply(
+            assign({}, mock.DEFAULT_ISMASTER, {
+              msg: 'isdbgrid',
+              logicalSessionTimeoutMinutes: 10
+            })
+          );
+        });
+
+        var mongos = new Mongos([{ host: 'localhost', port: 37019 }], {
+          connectionTimeout: 30000,
+          socketTimeout: 30000,
+          haInterval: 500,
+          size: 1
+        });
+
+        mongos.on('error', done);
+        mongos.once('connect', () => {
+          expect(mongos.logicalSessionTimeoutMinutes).to.equal(10);
+          mongos.destroy();
+          done();
+        });
+
+        mongos.connect();
+      });
+    }
+  });
+
+  it(
+    'should track `logicalSessionTimeoutMinutes` for replset topology, choosing the lowest value',
+    {
+      metadata: { requires: { topology: 'single' } },
+      test: function(done) {
+        var ReplSet = this.configuration.mongo.ReplSet;
+
+        const replSetState = new MockReplSetState();
+        replSetState.primary[0].logicalSessionTimeoutMinutes = 426;
+        replSetState.firstSecondary[0].logicalSessionTimeoutMinutes = 1;
+        replSetState.arbiter[0].logicalSessionTimeoutMinutes = 32;
+
+        co(function*() {
+          const primaryServer = yield mock.createServer(32000, 'localhost');
+          const firstSecondaryServer = yield mock.createServer(32001, 'localhost');
+          const arbiterServer = yield mock.createServer(32002, 'localhost');
+
+          primaryServer.setMessageHandler(request => {
+            var doc = request.document;
+            if (doc.ismaster) {
+              request.reply(replSetState.primary[0]);
+            }
+          });
+
+          firstSecondaryServer.setMessageHandler(request => {
+            var doc = request.document;
+            if (doc.ismaster) {
+              request.reply(replSetState.firstSecondary[0]);
+            }
+          });
+
+          arbiterServer.setMessageHandler(request => {
+            var doc = request.document;
+            if (doc.ismaster) {
+              request.reply(replSetState.arbiter[0]);
+            }
+          });
+
+          var replset = new ReplSet(
+            [{ host: '127.0.0.1', port: 32000 }, { host: '127.0.0.1', port: 32001 }],
+            {
+              setName: 'rs',
+              connectionTimeout: 3000,
+              socketTimeout: 0,
+              haInterval: 100,
+              size: 1
+            }
+          );
+
+          replset.on('error', done);
+          replset.once('connect', () => {
+            expect(replset.logicalSessionTimeoutMinutes).to.equal(1);
+            replset.destroy();
+            done();
+          });
+
+          replset.connect();
+        });
+      }
+    }
+  );
+
+  it('should set `logicalSessionTimeoutMinutes` to `null` if any incoming server is `null`', {
```

Comment:
this looks identical to previous test, minus the secondary's logicalSessionTimeoutMinutes value and the replSet's expected value. maybe you could combine them and just loop over values to use for the secondary and the expected value?

Reply:
Yeah, totally. I'm actually refactoring this presently for the cluster time gossiping stuff, I think we should do the refactor there instead.